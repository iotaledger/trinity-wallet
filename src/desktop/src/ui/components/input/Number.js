import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import Icon from 'ui/components/Icon';

import css from './input.scss';

/**
 * Single number input component
 */
export default class Number extends React.PureComponent {
    static propTypes = {
        /** Is input inline style */
        inline: PropTypes.bool,
        /** Is input disabled */
        disabled: PropTypes.bool,
        /** Number input label */
        label: PropTypes.string,
        /** Current input value */
        value: PropTypes.number.isRequired,
        /** Min valid value */
        min: PropTypes.number,
        /** Max valid value */
        max: PropTypes.number,
        /** Should input focus when changed to true */
        focus: PropTypes.bool,
        /** Value change event function
         * @param {number} value - Current input value
         */
        onChange: PropTypes.func.isRequired,
    };

    componentDidMount() {
        if (this.props.focus) {
            this.input.focus();
        }
    }

    componentWillReceiveProps(nextProps) {
        if (!this.props.focus && nextProps.focus) {
            this.input.focus();
        }
    }

    onInput(input) {
        const { min, max, onChange, value } = this.props;

        if (isNaN(input)) {
            input = value;
        }

        if (typeof min === 'number') {
            input = Math.max(input, min);
        }

        if (typeof max === 'number') {
            input = Math.min(input, max);
        }

        onChange(input);
    }

    render() {
        const { disabled, label, inline, value } = this.props;

        return (
            <div className={classNames(css.input, css.number, inline && css.inline, disabled && css.disabled)}>
                <fieldset>
                    <div>
                        <small>{label}</small>
                        <span onClick={() => this.onInput(value - 1)}>
                            <Icon icon="chevronLeft" size={inline ? 9 : 12} />
                        </span>
                        <input
                            ref={(input) => {
                                this.input = input;
                            }}
                            type="number"
                            value={value}
                            min="0"
                            onChange={(e) => this.onInput(parseInt(e.target.value))}
                        />
                        <span onClick={() => this.onInput(value + 1)}>
                            <Icon icon="chevronRight" size={inline ? 9 : 12} />
                        </span>
                    </div>
                </fieldset>
            </div>
        );
    }
}
