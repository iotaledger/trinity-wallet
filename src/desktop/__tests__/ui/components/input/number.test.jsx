import React from 'react';
import { shallow } from 'enzyme';

import Number from 'ui/components/input/Number';

const props = {
    value: 10,
    label: 'Foo',
    onChange: jest.fn(),
};

describe('TeNumberxt component', () => {
    test('Render the component', () => {
        const wrapper = shallow(<Number {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('Input value', () => {
        const wrapper = shallow(<Number {...props} />);

        expect(wrapper.find('input').props().value).toEqual(10);
    });

    test('Input change callback', () => {
        const wrapper = shallow(<Number {...props} />);

        wrapper.find('input').simulate('change', { target: { value: '30' } });
        expect(props.onChange).toHaveBeenLastCalledWith(30);
    });

    test('Input label', () => {
        const wrapper = shallow(<Number {...props} />);

        expect(wrapper.find('small').text()).toEqual('Foo');
    });

    test('Input decrement callback', () => {
        const wrapper = shallow(<Number {...props} />);

        wrapper
            .find('span')
            .first()
            .simulate('click');
        expect(props.onChange).toHaveBeenLastCalledWith(9);
    });

    test('Input decrement callback', () => {
        const wrapper = shallow(<Number {...props} />);

        wrapper
            .find('span')
            .last()
            .simulate('click');
        expect(props.onChange).toHaveBeenLastCalledWith(11);
    });
});
