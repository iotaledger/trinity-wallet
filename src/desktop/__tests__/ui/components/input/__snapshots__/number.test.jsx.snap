// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeNumberxt component Render the component 1`] = `
<div
  className="input number"
>
  <fieldset>
    <div>
      <small>
        Foo
      </small>
      <span
        onClick={[Function]}
      >
        <Icon
          icon="chevronLeft"
          size={12}
        />
      </span>
      <input
        min="0"
        onChange={[Function]}
        type="number"
        value={10}
      />
      <span
        onClick={[Function]}
      >
        <Icon
          icon="chevronRight"
          size={12}
        />
      </span>
    </div>
  </fieldset>
</div>
`;
